### ISSUE TITLE
Security of VNC console in qemu builder

### ISSUE THREAD

**Issue by candlerb** [mitchellh/packer#3533]
* Packer v0.10.1
* Ubuntu 14.04

The qemu builder runs qemu with flag `-vnc 0.0.0.0:<port>`

This means that anyone else on your network can connect to the console port. They can type commands into the boot screen, and in principle take over the build process, inject malware etc. This could be done over the Internet if you're running packer on a cloud and haven't blocked ports 5900-6000.

To close this hole, bind to 127.0.0.1 instead:

~~~
--- a/builder/qemu/step_run.go
+++ b/builder/qemu/step_run.go
@@ -66,7 +66,7 @@ func getCommandArgs(bootDrive string, state multistep.StateBag) ([]string, error
        ui := state.Get("ui").(packer.Ui)
        driver := state.Get("driver").(Driver)

-       vnc := fmt.Sprintf("0.0.0.0:%d", vncPort-5900)
+       vnc := fmt.Sprintf("127.0.0.1:%d", vncPort-5900)
        vmName := config.VMName
        imgPath := filepath.Join(config.OutputDir, vmName)

~~~

I have checked that packer does connect to 127.0.0.1 when typing the boot command:

~~~
$ sudo netstat -natp | egrep ':59[0-9][0-9]\b'
tcp        0      0 0.0.0.0:5918            0.0.0.0:*               LISTEN      22254/qemu-system-x
tcp        0      0 127.0.0.1:41171         127.0.0.1:5918          ESTABLISHED 22224/packer
tcp        0      0 127.0.0.1:5918          127.0.0.1:41171         ESTABLISHED 22254/qemu-system-x
~~~

However there also needs to be a way to override this, since I *do* want to be able to access VNC from a remote machine in a trusted environment, when running headless.

If `vnc_port` is chosen early enough to be put into `qemuArgsTemplateData`, it would allow you to configure:

~~~
"qemuargs": [
  [ "-vnc", "0.0.0.0:{{ .VNCPort }}" ]
]
~~~

Another option would be to generate a random password for the VNC console, and display it to stdout where packer is running - but this is less convenient to use, and I think more complicated than needed.

**Comment by vtolstov** [comment]
I'm prefer to generate random password, becasue i'm debug my builds from remote hosts and want to connect to it from my notebook

**Comment by candlerb** [comment]
If you want to be able to use existing JSON files without modifying them, then perhaps a command-line flag would be better. I'm fine with that, but I do think the user should explicitly request insecure operation one way or the other.

Aside: for insecure mode `-vnc [::]:nnnn` would be better, as this will allow both v4 and v6 connections.

Another option is to do ssh tunnelling, e.g.

    ssh -L 5959:127.0.0.1:5959 <remotehost>

but this isn't very convenient with the dynamic VNC port (unless you put 100 separate port-forwarding rules in `.ssh/config`, or use the SOCKS port-forwarding feature)

**Comment by vtolstov** [comment]
I know about port forwarding, but i don't think that we need create more troubles....
You can specify port ranges for vnc in packer json file and use iptables to block access from remote locations.

**Comment by cbednarski** [comment]
Thanks for the report. I think this has come up before and it's unlikely that a hard-coded setting here will satisfy everyone. If you'd like to make this configurable I'd be happy to review a PR for it! Options to specify the IP and port would probably satisfy most people, and after that we can change the default to a loopback interface.

**Comment by candlerb** [comment]
Made configurable in pull request #3559 - also added a security note explaining how it is open by default and how to close it.

If you prefer, I can change the default to 127.0.0.1, remove the security note, and instead explain under the 'headless' setting how to open it.

**Comment by candlerb** [comment]
Alternative patch #3570 changes the default to 127.0.0.1

**Comment by imduffy15** [comment]
This can be closed given the merging of #3574

**Comment by cbednarski** [comment]
Thanks for the reminder, @imduffy15!
